### PR TITLE
Atomic migrations

### PR DESCRIPTION
**Pull request by chrismwendt**
**tl;dr** this makes migrations atomic, which ensures that successful migrations always leave the DB in a clean state.

**Prior to this change**, it was possible for the frontend to disconnect from Postgres halfway through a migration (e.g. when the migration takes [>5 minutes and k8s kills the frontend because it's not responding](https://github.com/sourcegraph/deploy-sourcegraph-dot-com/blob/78d19bd4f013a7193ddcf3e8739086a922452eb5/base/frontend/sourcegraph-frontend.Deployment.yaml#L66)), leaving the DB marked as being in a dirty state despite the migration eventually succeeding.

**After this change**, each migration will set `dirty=false` within the transaction to guarantee that when a migration succeeds the DB will be in a clean state.

More context:

- Motivating prod error https://sourcegraph.slack.com/archives/C0J618TTM/p1579282575001900
- Investigation https://sourcegraph.slack.com/archives/C07KZF47K/p1579287343014500
- Filed issue https://github.com/golang-migrate/migrate/issues/325 (and submitted a PR https://github.com/golang-migrate/migrate/pull/326, probably won't get merged soon)
- Problem statement and alternatives https://sourcegraph.slack.com/archives/C07KZF47K/p1579546621063800

A few alternatives:

- Accept that migrations can fail in this way and document how to fix the DB when this happens
- Use my patched version of golang-migrate until a satisfactory solution has been upstreamed
- Mitigation: increase the timeout for DB migrations
- Don't merge slow migrations into master, somehow check perf before merging
- Workaround 1: SET version=..., dirty=false at the end of each migration in our migrations/ directory
- Workaround 2 (implemented in this PR): splice that statement into the migration at runtime by intercepting the migration loader
- Workaround 3: inline the ~40 lines of relevant code from golang-migrate and patch that in our source

3 teammates [supported](https://sourcegraph.slack.com/archives/C07KZF47K/p1579547161066000?thread_ts=1579546621.063800&cid=C07KZF47K) workaround 3:

> I vote for workaround 3, since we’re postgres only and the value-add from that lib is not that high. We can also do more postgres specific things if we go workaround 3 route and do stuff more tailored towards our exact needs.

Upon further consideration, it became more clear to me that [workaround 2 would be better](https://sourcegraph.slack.com/archives/C07KZF47K/p1579563038077500?thread_ts=1579546621.063800&cid=C07KZF47K):

> @eric and I chatted about this in depth and realized that workaround 3 is not very desirable because it would not play well with the `migrate` command line tool.
> 
> Workaround 2 is more compatible and doesn't change the control flow as much.
> 
> Given that, we plan to proceed with workaround 2.

This PR is based on https://github.com/sourcegraph/sourcegraph/pull/7905, which is expected to be merged before this PR.

### CODE REVIEW

**Comment by unknwon** [comment]
FYI, I changed the merge base branch to the `clean-up-migrate-db` for easier review (real change set for this PR):

image

Don't forget to change it back to `master` once #7905 is merged :)

File: internal/db/dbutil/dbutil.go [hunk 1]
```diff
@@ -111,8 +113,33 @@ func NewDB(dsn, app string) (*sql.DB, error) {
...
+// once the migration is committed.
+//
+// See https://github.com/golang-migrate/migrate/issues/325.
+func InjectVersionUpdate(f bindata.AssetFunc) bindata.AssetFunc {
+	return func(name string) ([]byte, error) {
+		oldContents, err := f(name)
+		if err != nil {
+			return nil, err
+		}
+		versionParts := strings.Split(path.Base(name), "_")
```
Comment (unknwon) [inline internal/db/dbutil/dbutil.go]: nit:
```suggestion
		versionParts := strings.SplitN(path.Base(name), "_", 2)
```

File: internal/db/dbutil/dbutil.go [hunk 2]
```diff
@@ -111,8 +113,33 @@ func NewDB(dsn, app string) (*sql.DB, error) {
...
+// InjectVersionUpdate fixes the dirty state (set by golang-migrate) after a
+// successful migration. If the frontend starts a migration that will turn out
+// to be successful but does not stay alive for the duration of the query due to
+// a startup timeout, there will be no chance to set the new version or unset
+// the dirty flag. This function ensures that each successful migration sets the
+// version and dirty flag itself, without requiring the frontend to be alive
+// once the migration is committed.
+//
+// See https://github.com/golang-migrate/migrate/issues/325.
+func InjectVersionUpdate(f bindata.AssetFunc) bindata.AssetFunc {
```
Comment (unknwon) [inline internal/db/dbutil/dbutil.go]: nit: Is there a reason to make this function exported? Uint tests are also seem worth adding.

File: internal/db/dbutil/dbutil.go [hunk 3]
```diff
@@ -111,8 +113,33 @@ func NewDB(dsn, app string) (*sql.DB, error) {
...
+	return func(name string) ([]byte, error) {
```
Comment (unknwon) [inline internal/db/dbutil/dbutil.go]: Ideally we could give this function a concrete name, e.g. `setVersionAfterEachMigrate`, and let `InjectVersionUpdate` accepts a list of "injectors" because just by looking at "InjectVersionUpdate" alone, it's seems too general. It's just an idea :)

File: internal/db/dbutil/dbutil.go [hunk 4]
```diff
@@ -111,8 +113,33 @@ func NewDB(dsn, app string) (*sql.DB, error) {
...
+		oldContents, err := f(name)
+		if err != nil {
+			return nil, err
+		}
+		versionParts := strings.Split(path.Base(name), "_")
+		if len(versionParts) < 2 {
+			return nil, errors.New("expected migration filename to be of the form <version>_<name>")
+		}
+		version := versionParts[0]
+		newContents := strings.Replace(string(oldContents), "COMMIT;", fmt.Sprintf("UPDATE schema_migrations SET version='%s', dirty=false;\nCOMMIT;", version), 1)
```
Comment (tsenart) [inline internal/db/dbutil/dbutil.go]: I would expect golang-migrate to already update the version in the database. If that premise is valid, why do we need to re-update it? Would it be enough to just set dirty=false (and avoid all the version parsing code)?

Comment (chrismwendt) [reply internal/db/dbutil/dbutil.go]: > Would it be enough to just set dirty=false (and avoid all the version parsing code)?

Oh right, yes, golang-migrate sets the version so we can drop the version parsing code.

Comment (efritz) [reply internal/db/dbutil/dbutil.go]: No you can't!

If you set dirty=false but not the version, the frontend will not be around to update the version (necessarily) so you'll have applied the migration but the version will be the old one. This will cause the next frontend attempt to fail (forever).

File: internal/db/dbutil/dbutil.go (hunk 2 above)
Comment (chrismwendt) [reply internal/db/dbutil/dbutil.go]: Made this function private and added a unit test.

File: internal/db/dbutil/dbutil.go (hunk 3 above)
Comment (chrismwendt) [reply internal/db/dbutil/dbutil.go]: Hmm, I played around with this but couldn't quickly come up with something that seemed simpler or more readable. We can always improve this later, too.

File: internal/db/dbutil/dbutil.go (hunk 4 above)
Comment (chrismwendt) [reply internal/db/dbutil/dbutil.go]: @efritz golang-migrate sets the version and dirty=true prior to the migration.

https://github.com/golang-migrate/migrate/blob/v4.8.0/migrate.go#L739-L754

```
			// set version with dirty state
			if err := m.databaseDrv.SetVersion(migr.TargetVersion, true); err != nil {
				return err
			}

			if migr.Body != nil {
				m.logVerbosePrintf("Read and execute %v\n", migr.LogString())
				if err := m.databaseDrv.Run(migr.BufferedBody); err != nil {
					return err
				}
			}

			// set clean state
			if err := m.databaseDrv.SetVersion(migr.TargetVersion, false); err != nil {
				return err
			}
```

I believe that means the frontend doesn't need to be around to update the version.

I won't merge this in until we reach consensus here.

Comment (chrismwendt) [reply internal/db/dbutil/dbutil.go]: @efritz golang-migrate sets the version and dirty=true prior to the migration.

https://github.com/golang-migrate/migrate/blob/v4.8.0/migrate.go#L739-L754

```
			// set version with dirty state
			if err := m.databaseDrv.SetVersion(migr.TargetVersion, true); err != nil {
				return err
			}

			if migr.Body != nil {
				m.logVerbosePrintf("Read and execute %v\n", migr.LogString())
				if err := m.databaseDrv.Run(migr.BufferedBody); err != nil {
					return err
				}
			}

			// set clean state
			if err := m.databaseDrv.SetVersion(migr.TargetVersion, false); err != nil {
				return err
			}
```

I believe that means the frontend doesn't need to be around to update the version.

I won't merge this in until we reach consensus here.

Comment (efritz) [reply internal/db/dbutil/dbutil.go]: Oh, I see. I thought it only set dirty=true first, then dirty=false, version=target at the end.

I'm convinced you can remove it now.

**Comment by chrismwendt** [comment]
I merged this into the wrong branch 🤦 It never made it into master.

CleanShot 2021-09-17 at 16 20 46@2x

I'm attempting this again in (TODO make PR)

**Comment by efritz** [comment]
@chrismwendt I think there's a problem in this solution as well. If the migration continues to run after the timeout (and the client disconnects), then it may be the case that a new frontend will try to re-attempt the migration and fail.

This could cause a lot of noise during long migrations that still result in a crashloopbackoff that might eventually resolve without user intervention.

Like, it's a better state than we're in, but it's not nearly as good as what [RFC 469](https://docs.google.com/document/d/1_wqvNbEGhMzu45CHZNrLQYys7gl0AtVpLluGp4cbQKk)'s proposal guarantees (imo).

**Comment by chrismwendt** [comment]
@efritz It looks like golang-migrate [acquires an advisory lock **before** checking the dirty state](https://github.com/golang-migrate/migrate/blob/10a92d0cc81953b21fae28c9becd4d594dcef868/migrate.go#L268-L272), so frontends won't see the dirty state:

CleanShot 2021-09-19 at 12 22 53@2x

If it did it the other way around, then on upgrade all but then frontend that first began the migration would throw errors about a dirty DB (which doesn't happen).

Although the frontend isn't around to explicitly release the lock, [Postgres automatically releases it](https://www.postgresql.org/docs/9.4/explicit-locking.html#ADVISORY-LOCKS) once the session ends (which I'm guessing is immediately after the SQL commands finish):

CleanShot 2021-09-19 at 12 42 41@2x

I tried interrupting migrations in a few different ways to see the behavior:


https://user-images.githubusercontent.com/1387653/133941078-8564dab4-043f-468d-be50-9fe593d2e521.mp4

It seems to me that `SET dirty = 'f'` within the migration produces desirable behavior, but it's been a long time since I first looked at this. Am I misunderstanding anything here?
